### PR TITLE
Darwin: Re-enable TAPI and change symbol visibility management

### PR DESCRIPTION
**Pull request by ksperling-apple**
Symbols are now hidden by default, but anything annotated with API availability (or deprecation, since that implies previous availability) is exported. Also add missing API availability to a number of classes that were introduced in the initial release.

### CODE REVIEW

File: src/darwin/Framework/Matter.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -2032,21 +2032,22 @@
 					"CHIP_ADDRESS_RESOLVE_IMPL_INCLUDE_HEADER=<lib/address_resolve/AddressResolve_DefaultImpl.h>",
 					"$(inherited)",
 				);
+				GCC_SYMBOLS_PRIVATE_EXTERN = YES;
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/Matter.xcodeproj/project.pbxproj:2035]: Hmm....  According to https://developer.apple.com/library/archive/documentation/DeveloperTools/Reference/XcodeBuildSettingRef/1-Build_Setting_Reference/build_setting_ref.html, "YES" is the default... but ok.

File: src/darwin/Framework/Matter.xcodeproj/project.pbxproj [hunk 2]
```diff
@@ -2032,21 +2032,22 @@
...
 				GCC_WARN_64_TO_32_BIT_CONVERSION = YES;
 				GCC_WARN_ABOUT_RETURN_TYPE = YES_ERROR;
 				GCC_WARN_UNDECLARED_SELECTOR = YES;
 				GCC_WARN_UNINITIALIZED_AUTOS = YES_AGGRESSIVE;
 				GCC_WARN_UNUSED_FUNCTION = YES;
 				GCC_WARN_UNUSED_VARIABLE = YES;
 				MACOSX_DEPLOYMENT_TARGET = 11.0;
-				MTL_ENABLE_DEBUG_INFO = INCLUDE_SOURCE;
-				MTL_FAST_MATH = YES;
```
Comment (bzbarsky-apple) [inline src/darwin/Framework/Matter.xcodeproj/project.pbxproj:2043]: How are the MTL_* changes related here?

**Review by bzbarsky-apple — APPROVED** [review]
It's great to see the missing availability annotations finally added!

File: src/darwin/Framework/Matter.xcodeproj/project.pbxproj (hunk 1 above)
Comment (ksperling-apple) [reply src/darwin/Framework/Matter.xcodeproj/project.pbxproj:2035]: I think that's an outdated document. According to how Xcode resolves this, the default is NO (and that's the behavior in practice.)

File: src/darwin/Framework/Matter.xcodeproj/project.pbxproj (hunk 2 above)
Comment (ksperling-apple) [reply src/darwin/Framework/Matter.xcodeproj/project.pbxproj:2043]: They're not related, just removing some unused "custom" settings that I think Xcode likes to add to new projects by default, but we're not using Metal so they're unneeded.

**Comment by ksperling-apple** [comment]
Hm the summary page claims the build succeeds but there is some test problem, but when I download the log archive I actually see these build errors:
```
2023-12-22T07:17:26.7579130Z In file included from /Users/runner/work/connectedhomeip/connectedhomeip/src/darwin/Framework/CHIP/MTRDeviceControllerStorageDelegate_Wrapper.h:24:
2023-12-22T07:17:26.7581310Z /Users/runner/work/connectedhomeip/connectedhomeip/src/darwin/Framework/CHIP/MTRDeviceControllerStorageDelegate.h:118:1: error: visibility does not match previous declaration
2023-12-22T07:17:26.7582960Z MTR_HIDDEN

2023-12-22T07:17:26.7626220Z In file included from /Users/runner/work/connectedhomeip/connectedhomeip/src/darwin/Framework/CHIP/MTRDeviceControllerParameters_Wrapper.h:24:
2023-12-22T07:17:26.7628570Z /Users/runner/work/connectedhomeip/connectedhomeip/src/darwin/Framework/CHIP/MTRDeviceControllerParameters.h:37:1: error: visibility does not match previous declaration
2023-12-22T07:17:26.7629890Z MTR_HIDDEN
```

`MTR_NEWLY_AVAILABLE` now implies MTR_EXPORT, so declaring it as MTR_HIDDEN also ends up erroring. I think we need to resolve the `MTR_PER_CONTROLLER_STORAGE_ENABLED` along with this (or before this lands).

**Comment by ksperling-apple** [comment]
Seems like TAPI is happy now, but now I'm seeing this error in the Darwin build that I don't get locally:
```
2024-01-24T04:14:57.0469110Z Testing failed:
2024-01-24T04:14:57.0473980Z    'MTRBaseDevice' is only available in macOS 13.0 or newer
2024-01-24T04:14:57.0475600Z    'MTRDeviceController' is only available in macOS 13.0 or newer
2024-01-24T04:14:57.0477150Z    'MTRDevice' is only available in macOS 13.0 or newer
2024-01-24T04:14:57.0478530Z    'MTRDevice' is only available in macOS 13.0 or newer
2024-01-24T04:14:57.0479870Z    'MTRDevice' is only available in macOS 13.0 or newer
2024-01-24T04:14:57.0480990Z    'MTRDevice' is only available in macOS 13.0 or newer
2024-01-24T04:14:57.0482170Z    'MTRDevice' is only available in macOS 13.0 or newer
2024-01-24T04:14:57.0483300Z    Testing cancelled because the build failed.
2024-01-24T04:14:57.0483950Z 
2024-01-24T04:14:57.0484160Z ** TEST FAILED **
2024-01-24T04:14:57.0485150Z 
2024-01-24T04:14:57.0485170Z 
2024-01-24T04:14:57.0485920Z The following build commands failed:
2024-01-24T04:14:57.0488870Z    SwiftEmitModule normal x86_64 Emitting\ module\ for\ MatterTests (in target 'MatterTests' from project 'Matter')
2024-01-24T04:14:57.0490810Z (1 failure)
```
Any ideas @bzbarsky-apple ?

**Pull request merged**